### PR TITLE
Fix `functional.max_poolNd` warning spam in the CI

### LINKED ISSUE #71257: `functional.max_poolNd` warnings are spamming the CI (~15000 times)
Warnings raised in `torch.nn.functional.max_poolNd` are being spammed in the CI. The count is around 15000, which makes it almost impossible to scroll through the log. These warnings were added in https://github.com/pytorch/pytorch/pull/62544.

Thanks to @Lezcano for reporting this issue, an example run is here: https://github.com/pytorch/pytorch/runs/4773322940?check_suite_focus=true (lines 39500  and up).

cc: @Lezcano @jbschlosser

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by krshrimali**
Fixes https://github.com/pytorch/pytorch/issues/71257.

Warnings have been removed, please see [this](https://github.com/pytorch/pytorch/pull/71258#issuecomment-1058503649) comment.

cc: @Lezcano @jbschlosser @zou3519

### CODE REVIEW

**Review by lezcano — APPROVED** [review]
Great, thanks!

cc @zou3519 to review and land, as he landed another warning fix for an autograd warning.

**Comment by krshrimali** [comment]
Looks like JIT is not happy with the changes, I have had no luck so far with this. Will try to take a look again, but will appreciate it if anyone has any hints for this JIT issue.

**Review by ngimel — CHANGES_REQUESTED** [review]
We should rethink deprecation strategy here, as we cannot just swap the order of 2 positional boolean args after some deprecation period, as this is extremely error-prone. IMO, we should first require them to be a kwarg-only argument, and then the order won't matter. 
While we are discussing what the correct deprecation strategy is, we should just remove these warnings. 
cc @jbschlosser, @mruberry

**Comment by jbschlosser** [comment]
Agree with @ngimel that making the args kw-only is a better way to go here. I don't see how we can move to kw-only without being immediately breaking, however.

Incidentally, these warnings are already raised only once in normal execution; I believe there is some test logic setting `warnings.simplefilter("always")` causing the CI spam.

**Comment by jbschlosser** [comment]
As discussed internally, we should throw an actionable warning that appears if users do not specify `ceil_mode` / `return_indices` as kwargs telling them that these args will be kw-only in the future and they should update their calls now to specify them as kwargs.

I believe `warnings.warn()` should still be fine to use for throwing a warning once, but we also need to fix the CI / test logic to avoid overwriting the single-throw behavior, resulting in spam.

**Comment by ngimel** [comment]
We should also fix callsites in the codebase so that kwagrs are used instead of positional args.

File: torch/nn/functional.py
Comment (IvanYashchuk) [inline torch/nn/functional.py:652]: There's `warnings.simplefilter('once', DeprecationWarning)` if you want your warning to be raised once. It's possible to wrap it in the context manager so that it doesn't affect any other warnings.

Comment (krshrimali) [reply torch/nn/functional.py:652]: Thanks, Ivan!

Comment (krshrimali) [reply torch/nn/functional.py]: Tried:

```python
warnings.simplefilter("once", category=DeprecationWarning)
    if ceil_mode != return_indices:
        warnings.warn("not equal", DeprecationWarning)
```

But it still spams in the tests, not sure why that would be!

Comment (Lezcano) [reply torch/nn/functional.py]: The previous approach was the correct one. `warnings` does not work well with this sort of warnings.

Comment (krshrimali) [reply torch/nn/functional.py]: The previous approach, unfortunately, didn't work well with JIT, I'll try again if there is a fix for that.

**Comment by Lezcano** [comment]
@albanD @eellison This PR is failing on some JIT-related issue. We are not quite sure what would be the correct approach to take here. Could any of you give you any pointers?

File: torch/nn/functional.py
Comment (albanD) [inline torch/nn/functional.py:710]: I think you don't want to store these on the function object. That's why jit is not happy.
Just make these proper globals: `_MAX_POOL1D_WITH_INDICES_WARN_COUNT = 0`

**Review by eellison — COMMENTED** [review]
I thought warnings already warn only once.. is this necessary ? Could we instead just wrap the particular spammy tests with 
```
with warnings.catch_warnings():
```

File: torch/nn/functional.py
Comment (krshrimali) [reply torch/nn/functional.py]: Thanks, @albanD for the suggestion. I updated the PR with the changes, but the tests fail again. I'm not really sure about this. cc: @eellison 

Also, using `with warnings.catch_warnings()` doesn't solve the spam. It still creates 1000s of warnings for `python3 test/test_ops.py -k max_pool<N>d`.

Comment (albanD) [reply torch/nn/functional.py]: Well, looks like jit doesn't support global either...
You can always guard the warning with a `if not torch.jit.is_scripting():` to hide that code from jit. People running script won't see the warning, but I think that's fine.

**Comment by jbschlosser** [comment]
I don't agree this is the right approach to take. I'd rather not see a hack in production logic to get around test logic setting `warnings.simplefilter("always")` somewhere. I'd like to reiterate Natalia's and my stance here:
* We should change the warning to be actionable: it should appear if users do not specify `ceil_mode` / `return_indices` as kwargs telling them that these args will be kw-only in the future and they should update their calls now to specify them as kwargs.
* We should also fix callsites in the codebase so that kwargs are used instead of positional args. This includes OpInfo sample generation.

With this done correctly, we won't generate warnings in the tests, removing the need to hack around the test logic.

**Comment by krshrimali** [comment]
> I don't agree this is the right approach to take. I'd rather not see a hack in production logic to get around test logic setting `warnings.simplefilter("always")` somewhere. I'd like to reiterate Natalia's and my stance here:
> 
> * We should change the warning to be actionable: it should appear if users do not specify `ceil_mode` / `return_indices` as kwargs telling them that these args will be kw-only in the future and they should update their calls now to specify them as kwargs.
> * We should also fix callsites in the codebase so that kwargs are used instead of positional args. This includes OpInfo sample generation.
> 
> With this done correctly, we won't generate warnings in the tests, removing the need to hack around the test logic.

Thanks, @jbschlosser for your inputs and the reminder :) I've attempted to do what you suggested and fixed the call sites as well, I'll appreciate if you can take a look whenever you find the time.

File: torch/nn/functional.py
Comment (krshrimali) [inline torch/nn/functional.py:849]: If these warnings are removed, the error for `Proxy object cannot be iterated` is not called. These test failures from `test/test_fx.py` will also be raised in the master branch if this branch of `if` is removed. Not sure of the solution, but the failures aren't directly related to the changes in this PR. Please correct me if I'm wrong.

cc: @jbschlosser

**Comment by krshrimali** [comment]
Hi, can anyone take a look at this PR? It's blocked on the comment here: https://github.com/pytorch/pytorch/pull/71258/files#r803511854. cc: @Lezcano @ngimel @jbschlosser

File: torch/nn/functional.py
Comment (jbschlosser) [reply torch/nn/functional.py:849]: @krshrimali You can fix the FX tests by editing the lines here:
https://github.com/pytorch/pytorch/blob/6c8e516a8034d7ab615025ac57b5bc2e7680efb1/test/test_fx.py#L3672-L3674

Changing `CONTROL_FLOW` -> `ARG_TYPE_MISMATCH` should do the trick. If not, there's a set of error messages the test is looking for here:
https://github.com/pytorch/pytorch/blob/6c8e516a8034d7ab615025ac57b5bc2e7680efb1/test/test_fx.py#L3566-L3573

**Comment by jbschlosser** [comment]
@krshrimali I added a comment that should unblock you - let me know if you run into more problems afterwards.

File: torch/nn/functional.py
Comment (krshrimali) [reply torch/nn/functional.py:849]: Thanks, @jbschlosser for your response and help! Please allow me a couple of days to get back to you and update the PR, if that's fine...

**Comment by Lezcano** [comment]
After a discussion with @ngimel and @jbschlosser , we figured out that there's no easy way to fix this at the moment due to limitations of TorchScript (it does not handle kwarg-only args, so we won't be able to change the position of make these arguments kwargs to be able to do a proper deprecation).

That being said, we considered that this small API inconsistency is a non-issue as, given that this function has 6 args, no one in their sane mind would specify all the args positionally. As such, we won't be fixing this issue.

**Comment by jbschlosser** [comment]
@Lezcano Given the decision not to move forward with a fix, we should either repurpose this PR or create a new one removing the warnings.

**Comment by Lezcano** [comment]
@krshrimali could you use this PR to remove the warnings? Thank you!

**Comment by krshrimali** [comment]
> 

Interesting, I would still say that it's an issue (as the ordering is still _wrong_), but I agree - one would rarely pass 6 args (FYI: PyTorch in their codebase had instances where 6 args were passed as positional :)..) + it is difficult to fix at this moment. I've attempted to make the changes required to revert back, and only remove warnings. In the PyTorch repo, wherever these args were passed as positional, I had converted them to keyword args. I hope that's okay.

File: test/test_fx.py
Comment (Lezcano) [inline test/test_fx.py:3706]: This is the only part I don't understand (not saying it's right or wrong, I just don't understand FX)

**Review by lezcano — APPROVED** [review]
This LGTM. The FX part I don't understand it, so I'll leave it for someone else to review.

File: test/test_fx.py
Comment (jbschlosser) [reply test/test_fx.py:3706]: These tests are a bit strange: symbolic tracing in FX is expected to fail for these operations, and the tests are just verifying the error message of the first problem encountered during tracing. If code changes, the first error tracing runs into may change, causing these tests to fail. The change here LGTM as long as the tests pass.

**Review by jbschlosser — APPROVED** [review]
LGTM - thanks for removing the warnings!

@ngimel want to sign off on this too or am I good to merge?

**Pull request closed without merging**
